### DISCUSSION TITLE
String copying optimization

### DISCUSSION THREAD

**Discussion by NN---** [dotnet/runtime#46361] [Ideas] [answered]
Can JIT optimize this case ?

```cs
string a = f();
string b = a + "\n"; // g(a + "\n")
// 'a' is not used after this line
```

It is obvious that sine `a` is not used later in the method.
Can JIT optimize copying the string here ?

**Comment by AlgorithmsAreCool** [comment]
I am unclear on what optimization you are asking for. What would the optimized result look like?

**Comment by NN---** [comment]
String string is immutable, string concatenation results in a new string and memory allocation.

**Reply by AlgorithmsAreCool** [reply]
Since `string` is well known to the runtime, the JIT could probably safely know that the concatenation is free from side effects and elide it if the result `b` is not used anywhere.

I don't think the JIT does this today.

**Reply by AlgorithmsAreCool** [reply]
According to sharplab, the JIT does not perform this optimization.

```
    L0000: push ebp
    L0001: mov ebp, esp
    L0003: call dword ptr [0x145ec6cc]
    L0009: mov ecx, eax
    L000b: mov edx, [0xbac95c4]
    L0011: call System.String.Concat(System.String, System.String) <-- not optimized away
    L0016: pop ebp
    L0017: ret
```

https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxNMAfAAgJgIwCwAUFgAwAEWeAdAEoCuAdjAJZhzUDCA9mAA4sANogDKiAG4sAxnCgBuYrlKKcAZnIBvYuR2UcOTdt3GqFCOQC85AGYAKAJQKixk3grBL5cwGpyAIgAdRj85cgB6MPIAc1sffyC/eyMXcMiAcgg08hYockZuGHJ6KDgAEy9reARyGAALHPJBFkY4ZPIAXza2gG0AWTg67lKASX5BW37BkbGAeT5WbkYoagA5bmHGJsZmqPsAXTasdSoANko3Gwc2rWcUygB2cgARCHgAFTYONYB3ajfuEQwBA7BxOYydIjtIA==

**Reply by NN---** [reply]
> 
> 
> Since `string` is well known to the runtime, the JIT could probably safely know that the concatenation is free from side effects and elide it if the result `b` is not used anywhere.
> 

This is exactly what I am asking :)
This optimization could be very useful.
There are a lot of such cases which can be optimized.

**Reply by AlgorithmsAreCool** [reply]
@EgorBo Does a lot of these small optimizations

**Reply by EgorBo** [reply]
I am not sure I understand the desirable behavior, `f()` allocates a string somehow than we need to append a char to it - it's a different object now, we can't avoid allocating `a` can we? is `f()` supposed to be inlined?

**Reply by AlgorithmsAreCool** [reply]
I think the `// 'a' is not used after this line` comment in the original snippet is a typo. The only optimization in can see here is to elide the call to String.Concat because it's result `b` is not used and Concat itself is a pure function.

**Reply by EgorBo** [reply]
Well, if `b` is not used the allocation indeed can be elided but it requires a (partial) escape analysis - it's not something JIT does today unfortunately

**Reply by acaly** [reply]
You probably want to add an overload in addition to `string f()` as `void f(StringBuilder)`, or use `Span<char>` if you have an idea on its maximum length. Then you can perform the concat yourself on your buffer. I believe that's how most C# code optimizes this kind of string concatenation today.

**Reply by NN---** [reply]
This option complicated the code which makes it harder to pass code review.

**Comment by danmoseley** [comment, marked as answer]
Answered - not possible without escape analysis.
